runtime-sdk/src/modules/rofl/app/client.rs

```diff
@@ -0,0 +1,262 @@
+        let enclave_rpc = RpcClient::new_runtime(
+            session::Builder::default()
+                .use_endorsement(true)
+                .quote_policy(Some(quote_policy))
+                .local_identity(identity)
+                .remote_enclaves(Some(enclaves)),
+            self.state.host.clone(),
+            ENCLAVE_RPC_ENDPOINT_RONL,
+            vec![],
+        );
+
+        let response: Vec<u8> = enclave_rpc
+            .secure_call(
+                METHOD_QUERY,
+                QueryRequest {
+                    round,
+                    method: method.to_string(),
+                    args: cbor::to_vec(args),
+                },
+            )
+            .await
+            .into_result()?;
+
+        Ok(cbor::from_slice(&response)?)
+    }
+
+    /// Securely perform gas estimation.
+    pub async fn estimate_gas(&self, req: EstimateGasQuery) -> Result<u64> {
+        let round = self.latest_round().await?;
+        self.query(round, "core.EstimateGas", req).await
+    }
+
+    /// Sign a given transaction and submit it.
+    ///
+    /// This method supports multiple transaction signers.
+    pub async fn multi_sign_and_submit_tx(
+        &self,
+        signers: &[&dyn Signer],
+        mut tx: transaction::Transaction,
+    ) -> Result<transaction::CallResult> {
+        let round = self.latest_round().await?;
+
+        // Resolve account nonces.
+        for signer in signers {
+            let sigspec = SignatureAddressSpec::try_from_pk(&signer.public_key())
+                .ok_or(anyhow!("signature scheme not supported"))?;
+            let address = Address::from_sigspec(&sigspec);
+            let nonce = self.account_nonce(round, address).await?;
+
+            tx.append_auth_signature(sigspec, nonce);
+
+            // If gas is not set, perform estimation.
+            if tx.fee_gas() == 0 {
+                let gas = self
+                    .estimate_gas(EstimateGasQuery {
+                        caller: Some(if let PublicKey::Secp256k1(pk) = signer.public_key() {
+                            CallerAddress::EthAddress(pk.to_eth_address().try_into().unwrap())
+                        } else {
+                            CallerAddress::Address(address)
+                        }),
+                        tx: tx.clone(),
+                        propagate_failures: false,
+                    })
+                    .await?;
+                tx.set_fee_gas(gas);
+            }
+        }
+
+        // Determine gas price. Currently we always use the native denomination.
+        let mgp = self.gas_price(round, token::Denomination::NATIVE).await?;
+        let fee = mgp.saturating_mul(tx.fee_gas().into());
+        tx.set_fee_amount(token::BaseUnits::new(fee, token::Denomination::NATIVE));
+
+        // Sign the transaction.
+        let mut tx = tx.prepare_for_signing();
+        for signer in signers {
+            tx.append_sign(*signer)?;
+        }
+        let tx = tx.finalize();
```

Comment:
Do we need so many steps? Since we are adding signatures one by one, the context is computed multiple times. To optimize, I would try to simplify this and merge 3 steps to `tx.sign(signers)`. No need to introduce `TransactionSigner`. We could also hide it in this case.

Finalize doesn't do anything. I would expect that at least checks if enough signatures are there (all required and threshold of multi sigs).

Reply:
This is modeled after the Go Client SDK API, I think it is nice if it is consistent?